File: python/2020/Day5.py
import sys
from common import DataAnalyzer



def personal_seat(data):
    data.sort()
    previous = data[0] - 1
    answer = -1

    for n in data:
        if previous != n - 1 and previous != n:
            answer = n - 1
        
        previous = n
    
    return answer


def locator(str, low, hi, upper, lower):
    x = hi
    n = low

    for c in str:
        if c == upper:
            x -= ((x - n + 1) / 2)
        elif c == lower:
            n += ((x - n + 1) / 2)

    return int(n)


def seat_decoder(data):
    ids = []

    for str in data:
        row = locator(str[:-3], 0, 127, 'F', 'B')
        col = locator(str[-3:], 0, 7, 'L', 'R')

        ids.append(row*8 + col)
    
    return ids


def second():
    data = DataAnalyzer.text("2020/day5.txt")
    ids = seat_decoder(data)

    print("(5.2) your seat id: {:}".format(personal_seat(ids)))


def first():
    data = DataAnalyzer.text("2020/day5.txt")
    print("(5.1) max seat id: {:d}".format(max(seat_decoder(data))))


def solve(puzzle):
    if puzzle == '1':
        first()
    elif puzzle == '2':
        second()
    else:
        first()
        second()


if __name__ == '__main__':
    solve(sys.argv[1])

File: js/2020/Day5.js
const {DataAnalyzer} = require('../common/DataAnalyzer.js')
let data = DataAnalyzer.strs("2020/day5.txt")


function checker(pass, lo, hi, start=0,end=127) {
    pass.split("").forEach(id => {
        if (id === lo) {
            end -= Math.floor((end-start+1)/2)
        } else {
            start += Math.floor((end-start+1)/2)
        }
    })

    return pass[pass.length-1] === hi ? end : start
}

function max(passes) {
    return Math.max(...passes)
}

function unique(passes) {
    let sorted = passes.sort((a, b) => a-b)
    let current = sorted[0]-1

    for (let i = 0; i < sorted.length; current = sorted[i], ++i) {
        if (sorted[i] !== current + 1) {
            return sorted[i]-1
        }
    }

    return -1
}

function calc(tabulator=max) {
    let passes = []
    data.forEach(pass => {
        row = checker(pass.slice(0, -3), 'F', 'B', 0, 127)
        col = checker(pass.slice(-3), 'L', 'R', 0, 7)
        passes.push((row*8)+col)
    })

    return tabulator(passes)
}

function one() {
    console.log("(2020.5.1): " + calc(max))
}

function two() {
    console.log("(2020.5.2): " + calc(unique))
}

switch(process.argv.slice(2)[0]) {
    case '1':
        one()
        break

    case '2':
        two()
        break
    
    default:
        one()
        two()
}